Generate the function's cache key lazily

ktosiek opened the pull request:
**Issue:** #2728 st.cache doesn't invalidate when internal function changes

**Description:**
* Moved the function hashing code out of `st.cache`.
* Moved the actual hashing into `wrapped_func`.

This way globals that are only created after the cached function can actually be included in the hash. Before this change only their names were included, which could lead to stale cache being used.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

ktosiek committed:
Generate the function's cache key lazily

This way globals that are only created after the cached function
can actually be included in the hash. Before this change only their
names were included, which could lead to stale cache being used.

ktosiek commented:
I'm not sure how I can write a test for this PR - are there any tests for how results change after file reload? Do I need to look at e2e tests?

kmcgrady commented on lib/streamlit/caching.py:
```diff
@@ -460,52 +460,7 @@ def cache(
+    cache_key = None
     _LOGGER.debug(
         "mem_cache key for %s.%s: %s", func.__module__, func.__qualname__, cache_key
     )
```
This logging message is less useful if `cache_key` is always `None` above

ktosiek replied:
Thank you, I've moved the logging message into `_hash_func`.

kmcgrady commented on lib/streamlit/caching.py:
```diff
@@ -528,6 +483,13 @@ def wrapped_func(*args, **kwargs):
+                # This way we can see complete globals, including functions
+                # defined after this one.
+                cache_key = _hash_func(func, hash_funcs)
+
```
Any reason we don't just do this before its called on line 563 (where it's used)?

ktosiek replied:
It's used just below, in a call to `_mem_caches.get_cache()` - this key is used to pick the right cache.

Actually, I'm not sure if adding this key to the `value_key` makes sense - there's no conflict, as the caches are separate.

kmcgrady reviewed (commented):
Hey @ktosiek Sorry for the delay in review. Had a couple thoughts/questions here. I am less familiar with the hashing in cache functionality, so I am ramping myself up here.

ktosiek committed:
Log function's cache_key after it's actually computed

kmcgrady commented:
Hey @ktosiek ! Thanks! I was confused at first (caching was not my expertise), but I'm getting the hang of it 😀. I took a look and the code is reasonable. As for a test, I found a place that might work!

[This test](https://github.com/streamlit/streamlit/blob/develop/lib/tests/streamlit/scriptrunner/script_runner_test.py#L368) verifies caches are preserved between script runs. You can create two scripts...and run the changed one after (instead of the same script). I'd verify that it failed as expected first based on the issue, and then see if your change correctly passes the test.

Let me know if I can help!

ktosiek committed:
Test invalidating st.cache() when a dependency changes

ktosiek commented:
I've added a test, and modified the existing `test_rerun_caching` - this way we can test both "nothing changed" and "some changes" for the same script.

I think this PR is now complete.

ktosiek committed:
Merge branch 'develop' of https://github.com/streamlit/streamlit into lazy-hashing-in-cache

 Conflicts:
	lib/tests/streamlit/scriptrunner/script_runner_test.py

kmcgrady commented:
@ktosiek Yes! That's it. My one only ask is that we have different text logged out (via `st.text`) in `cached_depending_on_not_yet_defined`. I ask because I want the test to be definitive that the correct cached function is rerunning and avoids any potential caching refactors.

Other than that, I am happy, we usually have a second person look at outside contributor code to verify, I'll get that done, so that we can merge this in early next week. Thank you for your patience!

ktosiek committed:
Make test_invalidating_cache more precise

kmcgrady reviewed (approved):
Awesome. This makes sense to me now. Going to follow up with our second reviewer.

vdonato commented on lib/streamlit/caching.py:
```diff
@@ -527,6 +479,13 @@ def wrapped_func(*args, **kwargs):
         def get_or_create_cached_value():
+            nonlocal cache_key
+            if cache_key is None:
+                # Wait with generating the cache key until the first call.
+                # This way we can see complete globals, including functions
+                # defined after this one.
```
Very minor thing, but it would be nice to mention here that if the hash is computed earlier, functions (and other globals) defined after the cached function will only have their names included in the hash.

ktosiek replied:
Would this be better?

                # Delay generating the cache key until the first call.
                # This way we can see values of globals, including functions
                # defined after this one.
                # If we generated the key earlier we would only hash those
                # globals by name, and miss changes in their code or value.

vdonato replied:
Yep, that's perfect, thanks! I think this is good to be merged once the comment is updated.

ktosiek committed:
Better explaination for generating cache_key late

kmcgrady commented:
Awesome work! Merging now!
